File: extensions/spark/kyuubi-spark-connector-tpcds/src/main/java/io/trino/tpcds/row/StoreRow.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.trino.tpcds.row;

import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_ADDRESS_CITY;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_ADDRESS_COUNTRY;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_ADDRESS_COUNTY;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_ADDRESS_GMT_OFFSET;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_ADDRESS_STATE;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_ADDRESS_STREET_NAME1;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_ADDRESS_STREET_NUM;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_ADDRESS_STREET_TYPE;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_ADDRESS_SUITE_NUM;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_ADDRESS_ZIP;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_CLOSED_DATE_ID;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_COMPANY_ID;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_COMPANY_NAME;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_DIVISION_ID;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_DIVISION_NAME;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_EMPLOYEES;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_FLOOR_SPACE;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_GEOGRAPHY_CLASS;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_HOURS;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_ID;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_MANAGER;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_MARKET_DESC;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_MARKET_ID;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_MARKET_MANAGER;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_NAME;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_REC_END_DATE_ID;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_REC_START_DATE_ID;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_SK;
import static io.trino.tpcds.generator.StoreGeneratorColumn.W_STORE_TAX_PERCENTAGE;
import static java.lang.String.format;

import io.trino.tpcds.type.Address;
import io.trino.tpcds.type.Decimal;
import org.apache.kyuubi.spark.connector.tpcds.row.KyuubiTPCDSTableRowWithNulls;

public class StoreRow extends KyuubiTPCDSTableRowWithNulls {

  private final long storeSk;
  private final String storeId;
  private final long recStartDateId;
  private final long recEndDateId;
  private final long closedDateId;
  private final String storeName;
  private final int employees;
  private final int floorSpace;
  private final String hours;
  private final String storeManager;
  private final int marketId;
  private final Decimal dTaxPercentage;
  private final String geographyClass;
  private final String marketDesc;
  private final String marketManager;
  private final long divisionId;
  private final String divisionName;
  private final long companyId;
  private final String companyName;
  private final Address address;

  public StoreRow(
      long nullBitMap,
      long storeSk,
      String storeId,
      long recStartDateId,
      long recEndDateId,
      long closedDateId,
      String storeName,
      int employees,
      int floorSpace,
      String hours,
      String storeManager,
      int marketId,
      Decimal dTaxPercentage,
      String geographyClass,
      String marketDesc,
      String marketManager,
      long divisionId,
      String divisionName,
      long companyId,
      String companyName,
      Address address) {
    super(nullBitMap, W_STORE_SK);
    this.storeSk = storeSk;
    this.storeId = storeId;
    this.recStartDateId = recStartDateId;
    this.recEndDateId = recEndDateId;
    this.closedDateId = closedDateId;
    this.storeName = storeName;
    this.employees = employees;
    this.floorSpace = floorSpace;
    this.hours = hours;
    this.storeManager = storeManager;
    this.marketId = marketId;
    this.dTaxPercentage = dTaxPercentage;
    this.geographyClass = geographyClass;
    this.marketDesc = marketDesc;
    this.marketManager = marketManager;
    this.divisionId = divisionId;
    this.divisionName = divisionName;
    this.companyId = companyId;
    this.companyName = companyName;
    this.address = address;
  }

  public long getStoreSk() {
    return storeSk;
  }

  public String getStoreId() {
    return storeId;
  }

  public long getRecStartDateId() {
    return recStartDateId;
  }

  public long getRecEndDateId() {
    return recEndDateId;
  }

  public String getGeographyClass() {
    return geographyClass;
  }

  public long getDivisionId() {
    return divisionId;
  }

  public String getDivisionName() {
    return divisionName;
  }

  public long getCompanyId() {
    return companyId;
  }

  public String getCompanyName() {
    return companyName;
  }

  public long getClosedDateId() {
    return closedDateId;
  }

  public String getStoreName() {
    return storeName;
  }

  public int getEmployees() {
    return employees;
  }

  public int getFloorSpace() {
    return floorSpace;
  }

  public String getHours() {
    return hours;
  }

  public String getStoreManager() {
    return storeManager;
  }

  public int getMarketId() {
    return marketId;
  }

  public Decimal getdTaxPercentage() {
    return dTaxPercentage;
  }

  public String getMarketDesc() {
    return marketDesc;
  }

  public String getMarketManager() {
    return marketManager;
  }

  public Address getAddress() {
    return address;
  }

  @Override
  public Object[] values() {
    return new Object[] {
      getOrNullForKey(storeSk, W_STORE_SK),
      getOrNull(storeId, W_STORE_ID),
      getDateOrNullFromJulianDays(recStartDateId, W_STORE_REC_START_DATE_ID),
      getDateOrNullFromJulianDays(recEndDateId, W_STORE_REC_END_DATE_ID),
      getOrNullForKey(closedDateId, W_STORE_CLOSED_DATE_ID),
      getOrNull(storeName, W_STORE_NAME),
      getOrNull(employees, W_STORE_EMPLOYEES),
      getOrNull(floorSpace, W_STORE_FLOOR_SPACE),
      getOrNull(hours, W_STORE_HOURS),
      getOrNull(storeManager, W_STORE_MANAGER),
      getOrNull(marketId, W_STORE_MARKET_ID),
      getOrNull(geographyClass, W_STORE_GEOGRAPHY_CLASS),
      getOrNull(marketDesc, W_STORE_MARKET_DESC),
      getOrNull(marketManager, W_STORE_MARKET_MANAGER),
      getOrNullForKey(divisionId, W_STORE_DIVISION_ID),
      getOrNull(divisionName, W_STORE_DIVISION_NAME),
      getOrNullForKey(companyId, W_STORE_COMPANY_ID),
      getOrNull(companyName, W_STORE_COMPANY_NAME),
      getOrNull(address.getStreetNumber(), W_STORE_ADDRESS_STREET_NUM),
      getOrNull(address.getStreetName(), W_STORE_ADDRESS_STREET_NAME1),
      getOrNull(address.getStreetType(), W_STORE_ADDRESS_STREET_TYPE),
      getOrNull(address.getSuiteNumber(), W_STORE_ADDRESS_SUITE_NUM),
      getOrNull(address.getCity(), W_STORE_ADDRESS_CITY),
      getOrNull(address.getCounty(), W_STORE_ADDRESS_COUNTY),
      getOrNull(address.getState(), W_STORE_ADDRESS_STATE),
      getOrNull(format("%05d", address.getZip()), W_STORE_ADDRESS_ZIP),
      getOrNull(address.getCountry(), W_STORE_ADDRESS_COUNTRY),
      getOrNull(address.getGmtOffset(), W_STORE_ADDRESS_GMT_OFFSET),
      getOrNull(dTaxPercentage, W_STORE_TAX_PERCENTAGE)
    };
  }
}

File: extensions/spark/kyuubi-spark-connector-tpcds/src/main/scala/io/trino/tpcds/row/StoreRow.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.trino.tpcds.row

import io.trino.tpcds.generator.StoreGeneratorColumn._
import io.trino.tpcds.`type`.Address
import io.trino.tpcds.`type`.Decimal
import java.util.{List => JList}

import org.apache.spark.sql.catalyst.expressions.GenericInternalRow
import org.apache.spark.sql.catalyst.InternalRow

class StoreRow(
                nullBitMap: Long,
                private val storeSk: Long,
                private val storeId: String,
                private val recStartDateId: Long,
                private val recEndDateId: Long,
                private val closedDateId: Long,
                private val storeName: String,
                private val employees: Int,
                private val floorSpace: Int,
                private val hours: String,
                private val storeManager: String,
                private val marketId: Int,
                private val dTaxPercentage: Decimal,
                private val geographyClass: String,
                private val marketDesc: String,
                private val marketManager: String,
                private val divisionId: Long,
                private val divisionName: String,
                private val companyId: Long,
                private val companyName: String,
                private val address: Address,
  ) extends TableRowWithNulls(nullBitMap, W_STORE_SK)
  with KyuubiTableRowWithNulls {

    override val nullBitMapInternal = nullBitMap
    override val firstColumnInternal = W_STORE_SK

  def getStoreSk: Long = storeSk
  def getStoreId: String = storeId
  def getRecStartDateId: Long = recStartDateId
  def getRecEndDateId: Long = recEndDateId
  def getClosedDateId: Long = closedDateId
  def getStoreName: String = storeName
  def getEmployees: Int = employees
  def getFloorSpace: Int = floorSpace
  def getHours: String = hours
  def getStoreManager: String = storeManager
  def getMarketId: Int = marketId
  def getDTaxPercentage: Decimal = dTaxPercentage
  def getGeographyClass: String = geographyClass
  def getMarketDesc: String = marketDesc
  def getMarketManager: String = marketManager
  def getDivisionId: Long = divisionId
  def getDivisionName: String = divisionName
  def getCompanyId: Long = companyId
  def getCompanyName: String = companyName
  def getAddress: Address = address

  override def getValues: JList[String] = throw new UnsupportedOperationException()

  def internalRow: InternalRow =
    new GenericInternalRow(
      Array(
    getLongOrNull(storeSk, W_STORE_SK),
    getStringOrNullInternal(storeId, W_STORE_ID),
    getDateOrNullFromJulianDays(recStartDateId, W_STORE_REC_START_DATE_ID),
    getDateOrNullFromJulianDays(recEndDateId, W_STORE_REC_END_DATE_ID),
    getLongOrNull(closedDateId, W_STORE_CLOSED_DATE_ID),
    getStringOrNullInternal(storeName, W_STORE_NAME),
    getIntOrNull(employees, W_STORE_EMPLOYEES),
    getIntOrNull(floorSpace, W_STORE_FLOOR_SPACE),
    getStringOrNullInternal(hours, W_STORE_HOURS),
    getStringOrNullInternal(storeManager, W_STORE_MANAGER),
    getIntOrNull(marketId, W_STORE_MARKET_ID),
    getStringOrNullInternal(geographyClass, W_STORE_GEOGRAPHY_CLASS),
    getStringOrNullInternal(marketDesc, W_STORE_MARKET_DESC),
    getStringOrNullInternal(marketManager, W_STORE_MARKET_MANAGER),
    getIntOrNull(divisionId, W_STORE_DIVISION_ID),
    getStringOrNullInternal(divisionName, W_STORE_DIVISION_NAME),
    getIntOrNull(companyId, W_STORE_COMPANY_ID),
    getStringOrNullInternal(companyName, W_STORE_COMPANY_NAME),
    getStringOrNullInternal(address.getStreetNumber.toString, W_STORE_ADDRESS_STREET_NUM),
    getStringOrNullInternal(address.getStreetName, W_STORE_ADDRESS_STREET_NAME1),
    getStringOrNullInternal(address.getStreetType, W_STORE_ADDRESS_STREET_TYPE),
    getStringOrNullInternal(address.getSuiteNumber, W_STORE_ADDRESS_SUITE_NUM),
    getStringOrNullInternal(address.getCity, W_STORE_ADDRESS_CITY),
    getStringOrNullInternal(address.getCounty, W_STORE_ADDRESS_COUNTY),
    getStringOrNullInternal(address.getState, W_STORE_ADDRESS_STATE),
    getStringOrNullInternal(f"${address.getZip}%05d", W_STORE_ADDRESS_ZIP),
    getStringOrNullInternal(address.getCountry, W_STORE_ADDRESS_COUNTRY),
    getDecimalOrNull(address.getGmtOffset, W_STORE_ADDRESS_GMT_OFFSET, 5, 2),
    getDecimalOrNull(dTaxPercentage, W_STORE_TAX_PERCENTAGE, 5, 2)))
}
